packages/@expo/cli/src/start/server/metro/createExpoStickyResolver.ts

```diff
@@ -0,0 +1,166 @@
+  });
+  return Object.fromEntries(
+    Object.entries(configResult.dependencies).map(([key, entry]) => [key, entry.root])
+  );
+};
+
+interface PlatformModuleDescription {
+  platform: AutolinkingPlatform;
+  moduleTestRe: RegExp;
+  resolvedModulePaths: Record<string, string>;
+}
+
+const getPlatformModuleDescription = async (
+  projectRoot: string,
+  platform: AutolinkingPlatform
+): Promise<PlatformModuleDescription> => {
+  const searchOptions = await getAutolinkingOptions(projectRoot, platform);
+  const autolinkingResolutions$ = getAutolinkingResolutions(searchOptions);
+  const reactNativeConfigResolutions$ = getReactNativeConfigResolutions(searchOptions);
+  const resolvedModulePaths = {
+    ...(await reactNativeConfigResolutions$),
+    ...(await autolinkingResolutions$),
+  };
+  const resolvedModuleNames = Object.keys(resolvedModulePaths);
+  debug(
+    `Sticky resolution for ${platform} registered ${resolvedModuleNames.length} resolutions:`,
+    resolvedModuleNames
+  );
+  return {
+    platform,
+    moduleTestRe: dependenciesToRegex(resolvedModuleNames),
+    resolvedModulePaths,
+  };
+};
+
+export type StickyModuleResolverInput = {
+  [platform in AutolinkingPlatform]?: PlatformModuleDescription;
+};
+
+export async function createStickyModuleResolverInput({
+  platforms,
+  projectRoot,
+}: {
+  projectRoot: string;
+  platforms: string[];
+}): Promise<StickyModuleResolverInput> {
+  return Object.fromEntries(
+    await Promise.all(
+      platforms
+        .filter((platform): platform is AutolinkingPlatform =>
+          AUTOLINKING_PLATFORMS.includes(platform as any)
+        )
+        .map(async (platform) => {
+          const platformModuleDescription = await getPlatformModuleDescription(
+            projectRoot,
+            platform
+          );
+          return [platformModuleDescription.platform, platformModuleDescription] as const;
+        })
+    )
+  ) as StickyModuleResolverInput;
+}
+
+export function createStickyModuleResolver(
+  input: StickyModuleResolverInput | undefined,
+  { getStrictResolver }: { getStrictResolver: StrictResolverFactory }
+): ExpoCustomMetroResolver | undefined {
+  if (!input) {
+    return undefined;
+  }
+
+  const fileSpecifierRe = /^[\\/]|^\.\.?(?:$|[\\/])/i;
+  const isAutolinkingPlatform = (platform: string | null): platform is AutolinkingPlatform =>
+    !!platform && (input as any)[platform] != null;
+
+  return function requestStickyModule(context, moduleName, platform) {
+    if (!isAutolinkingPlatform(platform)) {
+      return null;
+    } else if (fileSpecifierRe.test(moduleName)) {
+      return null;
+    }
+
+    const moduleDescription = input[platform]!;
+    const moduleMatch = moduleDescription.moduleTestRe.exec(moduleName);
+    if (moduleMatch) {
+      const resolvedModulePath =
+        moduleDescription.resolvedModulePaths[moduleMatch[1]] || moduleName;
+      const resolvedModuleRequest = resolvedModulePath + (moduleMatch[2] || '');
```

Comment:
I'm not entirely sure whether this breaks `package.json:exports` and ESM resolution. If so, we'll have to try something that's more similar to the fallback resolver's `getStrictResolver` call

Reply:
This is fixed with the latest two commits